diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/balancer/TestBalancer.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/balancer/TestBalancer.java
index 30a3a327d8c3..e844b7fa0895 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/balancer/TestBalancer.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/balancer/TestBalancer.java
@@ -44,12 +44,14 @@
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_NAMENODE_LAZY_PERSIST_FILE_SCRUB_INTERVAL_SEC;
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_WEB_AUTHENTICATION_KERBEROS_PRINCIPAL_KEY;
 import static org.apache.hadoop.test.PlatformAssumptions.assumeNotWindows;
+import static org.junit.Assert.*;
 
 import org.apache.hadoop.hdfs.protocol.ErasureCodingPolicy;
+import org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer;
+import org.apache.hadoop.hdfs.server.protocol.NamenodeProtocol;
+import org.apache.hadoop.ipc.Server;
+import org.apache.hadoop.ipc.metrics.RpcMetrics;
 import org.junit.AfterClass;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
 import java.io.File;
 import java.io.IOException;
@@ -59,15 +61,7 @@
 import java.net.InetSocketAddress;
 import java.net.URI;
 import java.security.PrivilegedExceptionAction;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Properties;
-import java.util.Random;
-import java.util.Set;
+import java.util.*;
 import java.util.concurrent.TimeoutException;
 
 import org.apache.commons.lang.StringUtils;
@@ -793,6 +787,7 @@ private void doTest(Configuration conf, long[] capacities, String[] racks,
    *   file and then later read from the file.
    * @throws Exception
    */
+
   private void doTest(Configuration conf, long[] capacities,
       String[] racks, long newCapacity, String newRack, NewNodeInfo nodes,
       boolean useTool, boolean useFile) throws Exception {
@@ -808,6 +803,12 @@ private void doTest(Configuration conf, long[] capacities,
                                 .racks(racks)
                                 .simulatedCapacities(capacities)
                                 .build();
+
+
+    NameNodeRpcServer rpcServer = (NameNodeRpcServer) cluster.getNameNodeRpc();
+    NNRPCCallQueueLengthMonitor monitor = new NNRPCCallQueueLengthMonitor(rpcServer.getClientRpcServer());
+    Thread monitorThread = new Thread(monitor);
+
     try {
       cluster.waitActive();
       client = NameNodeProxies.createProxy(conf, cluster.getFileSystem(0).getUri(),
@@ -884,6 +885,8 @@ private void doTest(Configuration conf, long[] capacities,
         }
       }
 
+      monitorThread.start();
+
       // run balancer and validate results
       if (useTool) {
         runBalancerCli(conf, totalUsedSpace, totalCapacity, p, useFile, expectedExcludedNodes);
@@ -892,9 +895,61 @@ private void doTest(Configuration conf, long[] capacities,
       }
     } finally {
       cluster.shutdown();
+      monitor.stop();
+      monitorThread.join();
+    }
+  }
+
+  @Test
+  public void testBalancerDispatcherSurge() throws Exception {
+    HdfsConfiguration conf = new HdfsConfiguration();
+    initConf(conf);
+    int numDataNodes = 150;
+    long[] capacities = new long[numDataNodes];
+    String[] racks = new String[numDataNodes];
+    Arrays.fill(capacities, CAPACITY);
+    Arrays.fill(racks, RACK0);
+    doTest(conf, capacities, racks, CAPACITY/2,
+            RACK0, false);
+  }
+
+  public class NNRPCCallQueueLengthMonitor implements Runnable {
+    Server rpcServer;
+    boolean isRunning;
+    List<Integer> nNRPCCallQueueLengths;
+
+    public NNRPCCallQueueLengthMonitor(Server rpcServer) {
+      this.rpcServer = rpcServer;
+      this.isRunning = false;
+      this.nNRPCCallQueueLengths = new LinkedList<>();
+    }
+
+    @Override
+    public void run() {
+      RpcMetrics metrics = rpcServer.getRpcMetrics();
+      isRunning = true;
+      try {
+        while (isRunning) {
+          nNRPCCallQueueLengths.add(Integer.valueOf(metrics.callQueueLength()));
+          Thread.sleep(5);
+        }
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+    }
+
+    public void stop () {
+      isRunning = false;
+      StringBuilder sb = new StringBuilder("NameNode RPC call queue length:");
+      for (int nNRPCCallQueueLength : nNRPCCallQueueLengths) {
+        sb.append(nNRPCCallQueueLength + " ");
+      }
+      sb.append("\n");
+      System.out.println(sb.toString());
     }
   }
 
+
   private void runBalancer(Configuration conf, long totalUsedSpace,
       long totalCapacity) throws Exception {
     runBalancer(conf, totalUsedSpace, totalCapacity,
